### PR TITLE
Adding UserStats plugin

### PR DESCRIPTION
**Pull request by avipatel91**
A plugin that gives statistics on message count has been added.

### CODE REVIEW

**Comment by avipatel91** [comment]
Added plugin for #59

File: src/plugins/UserStats.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            help: "Enter /UserStats to get statistics.",
+            type: Plugin.Type.PROXY,
+            needs: {
+                database: true
+            }
+        };
+    }
+
+    proxy(eventName, message) {
+        if (message.from.username) {
```
Comment (CapacitorSet) [inline src/plugins/UserStats.js]: Please use early returns for cleaner code:

``` js
if (!message.from.username) return;
const chatId = ...
```

File: src/plugins/UserStats.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                };
+            }
+            this.db["stat" + chatId][userId].messageCount++;
+            this.db["stat" + chatId].totalMessageCount++;
+        }
+        return Promise.resolve(message);
+    }
+
+    onText(message, reply) {
+        if (message.text === '/UserStats') {
```
Comment (CapacitorSet) [inline src/plugins/UserStats.js]: Please use [`Util.parseCommand`](https://github.com/crisbal/Telegram-Bot-Node/blob/es6/src/plugins/YouTube.js#L32):

``` js
import Util from "../Util";

/* ... */
const parts = Util.parseCommand(message.text, "userstats");
if (!parts) return;
```

File: src/plugins/UserStats.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            let text = `Total Messages:\n\n`;
+            const statsObject = this.db["stat" + message.chat.id];
+            const totalCount = statsObject.totalMessageCount;
+            for (let key in statsObject) {
+                if (key !== 'totalMessageCount') {
```
Comment (CapacitorSet) [inline src/plugins/UserStats.js]: This is also a good occasion to use early returns, except using `continue` of course.

``` js
for (const key in statsObject) {
    if (key === 'totalMessageCount') continue;
```

File: src/plugins/UserStats.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                    const username = statsObject[key].username;
+                    const id = statsObject[key].id;
+                    const count = statsObject[key].messageCount;
+                    const percentage = (count / totalCount * 100).toFixed(4);
+                    text += `${username} [${id}]: ${count} (${percentage}%)\n`;
+                }
+            }
+            reply({type: 'text', text: text});
```
Comment (CapacitorSet) [inline src/plugins/UserStats.js]: [ES6 syntactic sugar](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Property_definitions) is preferred: `reply({type: 'text', text})`

**Review by CapacitorSet — CHANGES_REQUESTED** [review]
Good job, it only needs some syntactic polishing.

**Comment by CapacitorSet** [comment]
Oh, one more thing, can you please sort the list (by number of messages descending) before printing it?

**Comment by avipatel91** [comment]
Thanks for the informative feedback! I applied the fixes for review.

File: src/plugins/UserStats.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            .map(item => {
+                return statsObject[item];
+            }).filter(item => {
+                return typeof item === "object";
+            }).sort((a, b) => {
+                return b.messageCount - a.messageCount;
+            });
+        for (let user of userList) {
+            if (user) {
+                const percentage = (user.messageCount / totalCount * 100).toFixed(4);
```
Comment (avipatel91) [inline src/plugins/UserStats.js]: Lines 53 and 54 surpass 80 character limit. Is there a better way to write it or break it apart?

File: src/plugins/UserStats.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                text += `${user.username} [${user.userId}]: ${user.messageCount} (${percentage}%)\n`;
```
Comment (avipatel91) [inline src/plugins/UserStats.js]: Lines 53 and 54 surpass 80 character limit. Is there a better way to write it or break it apart? @CapacitorSet

Comment (CapacitorSet) [reply src/plugins/UserStats.js]: That's not an issue. In fact, a few hours ago crisbal and I decided to ignore the rule and remove it from the linting rules (commit 190970a14bc7843858875d3cb295327443c6375e).

**Comment by CapacitorSet** [comment]
Please test your changes before applying them. As it is now, the plugin prints this:

```
Me: /userstats
Bot: Total Messages:

blackdev1l [undefined]: 141 (41.4706%)
crisbal [undefined]: 37 (10.8824%)
rTARS [undefined]: 2 (0.5882%)
```

I'll suggest changes using the review feature.

File: src/plugins/UserStats.js [hunk 7]
```diff
@@ -14,41 +15,45 @@ export default class UserStats extends Plugin {
...
```
Comment (CapacitorSet) [inline src/plugins/UserStats.js]: `user.userId` is undefined, at least on my setup. `user.id` is what you're looking for.

File: src/plugins/UserStats.js [hunk 8]
```diff
@@ -14,41 +15,45 @@ export default class UserStats extends Plugin {
...
+        const userList = Object.keys(statsObject)
+            .map(item => {
+                return statsObject[item];
+            }).filter(item => {
+                return typeof item === "object";
+            }).sort((a, b) => {
+                return b.messageCount - a.messageCount;
+            });
+        for (let user of userList) {
+            if (user) {
```
Comment (CapacitorSet) [inline src/plugins/UserStats.js]: Why is this line needed?

File: src/plugins/UserStats.js [hunk 9]
```diff
@@ -14,41 +15,45 @@ export default class UserStats extends Plugin {
...
-                    const percentage = (count / totalCount * 100).toFixed(4);
-                    text += `${username} [${id}]: ${count} (${percentage}%)\n`;
-                }
+        const parts = Util.parseCommand(message.text, "userstats");
+        if (!parts) return;
+        let text = `Total Messages:\n\n`;
+        const statsObject = this.db["stat" + message.chat.id];
+        const totalCount = statsObject.totalMessageCount;
+        const userList = Object.keys(statsObject)
+            .map(item => {
```
Comment (CapacitorSet) [inline src/plugins/UserStats.js]: Very nice use of map/filter! One suggestion, use the shorthand arrow notation:

``` js
.map(item => statsObject[item])
.filter(item => typeof item === "object")
.sort((a, b) => b.messageCount - a.messageCount)
```

**Review by CapacitorSet — CHANGES_REQUESTED** [review]
Please test your changes before applying them. As it is now, the plugin prints this:

```
Me: /userstats
Bot: Total Messages:

blackdev1l [undefined]: 141 (41.4706%)
crisbal [undefined]: 37 (10.8824%)
rTARS [undefined]: 2 (0.5882%)
```

File: src/plugins/UserStats.js (hunk 9 above)
Comment (avipatel91) [reply src/plugins/UserStats.js]: neat. clean and concise.

File: src/plugins/UserStats.js (hunk 8 above)
Comment (avipatel91) [reply src/plugins/UserStats.js]: Good catch. It was there when I was iterating over object for this different approach. Removing in next commit.

File: src/plugins/UserStats.js (hunk 7 above)
Comment (avipatel91) [reply src/plugins/UserStats.js]: You most likely have to delete plugin_UserStats in redis db since it is using object from before. I applied the syntactic sugar you suggested in different part of the code to https://github.com/avipatel91/Telegram-Bot-Node/blob/stats/src/plugins/UserStats.js#L27. id has changed to userId. Is that alright?

Comment (CapacitorSet) [reply src/plugins/UserStats.js]: Oh, yeah, it works now. Great!

**Review by CapacitorSet — APPROVED** [review]

**Comment by CapacitorSet** [comment]
Nice, I'll merge as soon as the Travis build finishes.
